### PR TITLE
Gradle project is created with deprecated features (Project.getConvention)

### LINKED ISSUE #1420: Gradle project is created with deprecated features (`Project.getConvention`)
Hi folks,

I noticed that creating a new project with Spring Boot 3.1.x + Gradle results in the following:

```
plugins {
	id 'java'
	id 'org.springframework.boot' version '3.1.0'
	id 'io.spring.dependency-management' version '1.1.0'
}

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '17'

repositories {
	mavenCentral()
}
```

Using `sourceCompatibility` is deprecated and has been marked for removal in Gradle 9.x. When a new project is created and using Gradle 8.2-rc-1, you will see the following:

```
The org.gradle.api.plugins.JavaPluginConvention type has been deprecated. This is scheduled to be removed in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/8.2-rc-1/userguide/upgrading_version_8.html#java_convention_deprecation
        at build_50ds3av89ty6gnq6l29fxeb5a.run(/something/sb-demo/build.gradle:9)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
Build file '/Users/rperezalcolea/Downloads/sb-demo/build.gradle': line 9
The org.gradle.api.plugins.Convention type has been deprecated. This is scheduled to be removed in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/8.2-rc-1/userguide/upgrading_version_8.html#deprecated_access_to_conventions
        at build_50ds3av89ty6gnq6l29fxeb5a.run(/something/sb-demo/build.gradle:9)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
```

The suggested way of doing this would be

```
java {
   sourceCompatibility = '17'
}
```

Happy to make this change but need to dig into how that piece is generated

### PR DESCRIPTION
**Pull request by prithvitewatia**
This PR closes #1420. In this PR gradle build project generation convention is updated to use Java configuration instead of `sourceCompatibility` property as they are deprecated and scheduled for removal in gradle 9.0. 
[Upgrading source compatibility](https://docs.gradle.org/8.2-rc-1/userguide/upgrading_version_8.html#java_convention_deprecation).

As so many test resources files had been updated to validate project generation according, I have tried to make small logical units of commits that may be helpful in reviewing this PR:

1. [Commit 1](https://github.com/prithvitewatia/initializr/commit/b0f5aed37c115c632297069cd8587af834111067). In this commit, I have simply updated .gen files to use new java configuration that is common for both groovy and kotlin.
2. [Commit 2](https://github.com/prithvitewatia/initializr/commit/045ee565db3926794b6598f1dd768dbcd1d0fb0c) In this commit, I have added a method to write java configuration instead of source compatibility.
3. [Commit 3](https://github.com/prithvitewatia/initializr/commit/98614b4b80bf352f040b6c8fc4cfe3e3b253f88b) In this commit, I have added a method that returns a string representation for java configuration to be used in tests.
4. [Commit 4](https://github.com/prithvitewatia/initializr/commit/e0f652a292a42e73a2c070a18f445ac3a5ea24f9) In this commit, I have updated the source compatibility test for Kotlin Dsl to check for java configuration.
5. [Commit 5](https://github.com/prithvitewatia/initializr/commit/321d1f821763bc7e110d83c95b8de1b6487af9c8) In this commit, I have updated the source compatibility test for groovy dsl to check for java configuration and parameterised this test.
6. [Commit 6](https://github.com/prithvitewatia/initializr/commit/1be85a06b640f645412212c4033814eb3a0978e1) In this method I have updated method to check for java source compatibility
7. [Commit 7](https://github.com/prithvitewatia/initializr/commit/1692dad186fb410debead4b90ad1773d660e16dd) In this 
commit, I have removed method that used to write source compatibility. Also created a class that will provide java version constant.
8. [Commit 8](https://github.com/prithvitewatia/initializr/commit/a74fc4275601a560b947cadc54b2dadd763710fa) Updated tests to check for Java configuration
9. [Commit 9](https://github.com/prithvitewatia/initializr/commit/ede1b9bba5a8d5135b3ad990ee55933f6186029e) Added java configuration in gradle build settings.

### CODE REVIEW

File: initializr-generator-spring/src/test/resources/project/gradle/annotation-processor-dependency-build.gradle.gen [hunk 1]
```diff
@@ -6,7 +6,9 @@ plugins {
 
 group = 'com.example'
 version = '0.0.1-SNAPSHOT'
-sourceCompatibility = '1.8'
+java {
```
Comment (snicoll) [inline initializr-generator-spring/src/test/resources/project/gradle/annotation-processor-dependency-build.gradle.gen:9]: That's an odd place to put this. Now that it isn't an attribute anymore, it should go after the dependencies IMO.

**Review by snicoll — COMMENTED** [review]
Thanks very much for the PR, I've added a comment for your consideration.

**Comment by snicoll** [comment]
We should consider also https://github.com/spring-io/initializr/issues/211 as part of this as we could simply remove the code for Kotlin-based projects.

File: initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java [hunk 2]
```diff
@@ -118,7 +118,9 @@ void buildDotGradleIsContributedWhenGeneratingGradleProject() {
 				"",
 				"group = 'com.example'",
 				"version = '0.0.1-SNAPSHOT'",
-				"sourceCompatibility = '11'",
+				"java {",
+				"    sourceCompatibility = JavaVersion.VERSION_11",
```
Comment (wilkinsona) [inline initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java:122]: I prefer that we continue to use a String for the version as it's more concise and, IMO, it reads better.

File: initializr-generator-spring/src/test/resources/project/gradle/annotation-processor-dependency-build.gradle.gen (hunk 1 above)
Comment (snicoll) [reply initializr-generator-spring/src/test/resources/project/gradle/annotation-processor-dependency-build.gradle.gen:9]: Thanks, I guess I am biased with how things work with Maven. It's all good keeping it there and yes with a blank line.

File: initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java (hunk 2 above)
Comment (prithvitewatia) [reply initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java:122]: @snicoll @wilkinsona, I read in https://docs.gradle.org/8.2-rc-1/dsl/org.gradle.api.plugins.JavaPluginExtension.html#org.gradle.api.plugins.JavaPluginExtension:sourceCompatibility
that source compatibility is going to be like java version constant. That is why I added it. So should we continue with String for version or add Java version constant ?

**Comment by prithvitewatia** [comment]
@snicoll @wilkinsona Thanks for the review. I have one question. Currently in [Gradle Build Settings](https://github.com/prithvitewatia/initializr/commit/ede1b9bba5a8d5135b3ad990ee55933f6186029e#diff-a0e95611961daa1ed5e9d81daf452abaff8d5d38587ea831decf1512fe54f5b6) I have added source compatibility as this is the only field we used to generate. However many other options are possible for java configuration as mentioned [here](https://docs.gradle.org/8.1/dsl/org.gradle.api.plugins.JavaPluginExtension.html#org.gradle.api.plugins.JavaPluginExtension:sourceCompatibility). Should I also add them or continue with source compatibility ?

File: initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java (hunk 2 above)
Comment (wilkinsona) [reply initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java:122]: Gradle will perform the `Object -> JavaVersion` conversion as it's https://docs.gradle.org/8.2-rc-1/javadoc/org/gradle/api/plugins/JavaPluginExtension.html#setSourceCompatibility-java.lang.Object- that will be called.

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/gradle/GradleBuildSettings.java [hunk 3]
```diff
@@ -131,4 +131,34 @@ public Dependency getDependency() {
 
 	}
 
+	/**
+	 * Class to map common configuration for Java based projects.
+	 *
+	 */
+	public static class JavaConfiguration {
```
[the remark this answers is not in the document]
Comment (snicoll) [reply initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/gradle/GradleBuildSettings.java:138]: Yeah, I agree that it's a bit early to introduce a new API for this.

[the remark this answers is not in the document]
Comment (prithvitewatia) [reply initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/gradle/GradleBuildSettings.java:138]: Thanks for the clarification. I will update the PR accordingly

File: initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java (hunk 2 above)
Comment (prithvitewatia) [reply initializr-generator-spring/src/test/java/io/spring/initializr/generator/spring/build/gradle/GradleProjectGenerationConfigurationTests.java:122]: Thanks. I understand now.

**Comment by snicoll** [comment]
@prithvitewatia I don't think that works. There is a difference between the Kotlin DSL and Kotlin as a language chosen for the application. We no longer generate a `sourceCompatibility` when the Kotlin DSL is chosen, whereas it should have been when Kotlin is chosen as a language.

Anyway, I don't like the idea of mixing two changes in the same PR. If you want us to consider this PR, please only focus on changing how the source compatibility is written, and ignoring #211 for now. Let me know if you'd like to review the PR or if you want me to take over.

**Comment by prithvitewatia** [comment]
@snicoll I might have intermixed two things. For now, I am planning to do changes only for #1423 in a new branch to better track changes.

**Pull request closed without merging**